### PR TITLE
Show positive/negative words in the search results depending on the sign of the sentiment score.  

### PR DESCRIPTION
**Pull request by ntarn**
If the search result has a positive score, it will highlight positive words as green, vice versa red for negative, but only if the Node.js sentiment module provides a list

### CODE REVIEW

File: search-sentiments-map/routes/sentiment-words.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+// the License.
+
+/** 
+ * Server-side script that uses the Node.js Sentiment module to get words that are 
+ * marked as negative or positive using the AFINN-165 wordlist.
+ */
+const express = require('express');
+const router = express.Router();  // Using Router to divide the app into modules.
+const sentimentApi = require('sentiment');
+var sentimentWords = new sentimentApi();
```
Comment (ntarn) [inline search-sentiments-map/routes/sentiment-words.js:22]: As a heads up: Unlike the other packages/modules we used, where you just use the constant directly I had to make a variable of the sentimentApi to call .analyze from the api. Just calling sentimentApi.analyze, would give the error of an unknown function. The documentation does it in the same way

File: search-sentiments-map/public/scripts/modal-script.js [hunk 2]
```diff
@@ -71,10 +71,11 @@ function toggleDisplay(id) {
 }
 
 /** 
- * Displays the top results in a country for current search trend on modal. 
+ * Displays the top results in a country for current search trend on modal
+ * and show positive/negative words depending on the sign of the sentiment score.  
  * @param {string} countryCode Two letter country code for selected country.
  */
-function displayTopResultsForCurrentTrend(countryCode) {
+async function displayTopResultsForCurrentTrend(countryCode) {
```
Comment (ChenyuHeidiZhang) [inline search-sentiments-map/public/scripts/modal-script.js:78]: This function seems too long. Consider dividing it up.

File: search-sentiments-map/public/scripts/modal-script.js [hunk 3]
```diff
@@ -91,18 +92,48 @@ function displayTopResultsForCurrentTrend(countryCode) {
...
+          .then(resultsJsonArray => resultsJsonArray.json())
+          .then(sentimentWordsResult => {
+            // i + 1 shows the index for each search result.
+            resultElement.innerHTML += (i + 1).toString() + '. ' + '<a href=' + 
+                results[i].link + '>' + results[i].htmlTitle + '</a><br>';
+            // Check to make sure that there is detection of positive and 
+            // negative words from the Node.js sentiment API.
+            if (sentimentWordsResult != null && 
+                sentimentWordsResult.positive.length != 0) {
+              var snippetArray = results[i].snippet.split(" ");
```
Comment (ChenyuHeidiZhang) [inline search-sentiments-map/public/scripts/modal-script.js:109]:
```suggestion
              var snippetArray = results[i].snippet.split(' ');
```

File: search-sentiments-map/public/scripts/modal-script.js [hunk 4]
```diff
@@ -91,18 +92,48 @@ function displayTopResultsForCurrentTrend(countryCode) {
...
+              snippetArray.forEach((word) => {
+                // Replace quotations or other symbols that may be attached to 
+                // the word.  
+                let checkWordInArray = word.replace(/[^a-z0-9+]+/gi, '');
+                if (results[i].score > 0 && sentimentWordsResult.positive
+                    .includes(checkWordInArray)) {
+                  resultElement.innerHTML += '<span style=\'color: ' + 
+                      POSITIVE_COLOR + ';\'>' + word + ' </span>';
```
Comment (ChenyuHeidiZhang) [inline search-sentiments-map/public/scripts/modal-script.js:117]: Use double quotes in HTML and you won't need the back slashes

File: search-sentiments-map/public/scripts/modal-script.js [hunk 5]
```diff
@@ -91,18 +92,48 @@ function displayTopResultsForCurrentTrend(countryCode) {
...
+                } else if (results[i].score < 0 && sentimentWordsResult.negative
+                    .includes(checkWordInArray)) {
+                  resultElement.innerHTML += '<span style=\'color: ' + 
+                      NEGATIVE_COLOR + ';\'>' + word + ' </span>';
```
Comment (ChenyuHeidiZhang) [inline search-sentiments-map/public/scripts/modal-script.js:121]: same above

File: search-sentiments-map/public/scripts/modal-script.js [hunk 6]
```diff
@@ -91,18 +92,48 @@ function displayTopResultsForCurrentTrend(countryCode) {
...
+                } else {
+                  resultElement.innerHTML += word + ' ';
+                }
+              });
+              resultElement.innerHTML += '<br><i>Sentiment Score: ' + 
+                  results[i].score.toFixed(1) + '</i><br>';
```
Comment (ChenyuHeidiZhang) [inline search-sentiments-map/public/scripts/modal-script.js:127]: Consider moving this out of the if-else block, because it is shared by both the if and else clauses

File: search-sentiments-map/public/scripts/modal-script.js [hunk 7]
```diff
@@ -91,18 +92,48 @@ function displayTopResultsForCurrentTrend(countryCode) {
         countryData[0].interest + '</b><br>';
     resultElement.innerHTML += '<b>Average Sentiment Score: ' + 
         countryData[0].averageSentiment.toFixed(1) + '</b><br>';
-
+
```
Comment (ChenyuHeidiZhang) [inline search-sentiments-map/public/scripts/modal-script.js:95]:
```suggestion

```

File: search-sentiments-map/routes/sentiment-words.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  getPositiveNegativeWords(titleSnippet).then(sentimentWordsJsonArray => {
+    res.setHeader('Content-Type', 'application/json');
+    res.send(sentimentWordsJsonArray);
+  });
+});
+
+async function getPositiveNegativeWords(titleSnippet) {
+  try {
+    var result = await sentimentWords.analyze(titleSnippet);
+    return result;
```
Comment (ChenyuHeidiZhang) [inline search-sentiments-map/routes/sentiment-words.js:34]:
```suggestion
    return await sentimentWords.analyze(titleSnippet);
```

File: search-sentiments-map/routes/sentiment-words.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  } catch (err) {
+    console.log(err);
+  }
+}
+
+module.exports.router = router;
+module.exports.getPositiveNegativeWords = getPositiveNegativeWords;
```
Comment (ChenyuHeidiZhang) [inline search-sentiments-map/routes/sentiment-words.js:41]: You don't need to export this function unless you are using it in some other script

File: search-sentiments-map/package.json [hunk 10]
```diff
@@ -13,7 +13,8 @@
     "express": "^4.17.1",
     "google-trends-api": "^4.9.0",
     "promise.allsettled": "^1.0.2",
-    "request": "^2.81.0"
+    "request": "^2.81.0",
+    "sentiment": "^5.0.2"
```
Comment (carmenBen) [inline search-sentiments-map/package.json:17]:
```suggestion
    "sentiment": "^5.0.2",
```

File: search-sentiments-map/public/scripts/modal-script.js [hunk 11]
```diff
@@ -71,10 +71,11 @@ function toggleDisplay(id) {
 }
 
 /** 
- * Displays the top results in a country for current search trend on modal. 
+ * Displays the top results in a country for current search trend on modal
+ * and show positive/negative words depending on the sign of the sentiment score.
```
Comment (carmenBen) [inline search-sentiments-map/public/scripts/modal-script.js:75]: Check line wrapping.
```suggestion
 * and show positive/negative words if they match the sign of the sentiment score.  
```

File: search-sentiments-map/public/scripts/modal-script.js [hunk 12]
```diff
@@ -71,10 +71,11 @@ function toggleDisplay(id) {
 }
 
 /** 
- * Displays the top results in a country for current search trend on modal. 
+ * Displays the top results in a country for current search trend on modal
```
Comment (carmenBen) [inline search-sentiments-map/public/scripts/modal-script.js:74]:
```suggestion
 * Displays the top results in a country for the current search trend on modal
```

File: search-sentiments-map/public/scripts/modal-script.js [hunk 13]
```diff
@@ -91,18 +92,48 @@ function displayTopResultsForCurrentTrend(countryCode) {
...
+      await fetch('/sentiment-words/' + results[i].title + results[i].snippet)
+          .then(resultsJsonArray => resultsJsonArray.json())
+          .then(sentimentWordsResult => {
+            // i + 1 shows the index for each search result.
+            resultElement.innerHTML += (i + 1).toString() + '. ' + '<a href=' + 
+                results[i].link + '>' + results[i].htmlTitle + '</a><br>';
+            // Check to make sure that there is detection of positive and 
+            // negative words from the Node.js sentiment API.
+            if (sentimentWordsResult != null && 
+                sentimentWordsResult.positive.length != 0) {
```
Comment (carmenBen) [inline search-sentiments-map/public/scripts/modal-script.js:108]:
```suggestion
                sentimentWordsResult.positive.length !== 0) {
```

Comment (carmenBen) [reply search-sentiments-map/public/scripts/modal-script.js:108]: Why do you only check that the positive length isn't 0, what about the negative length? Add a comment explaining why.

File: search-sentiments-map/public/scripts/modal-script.js [hunk 14]
```diff
@@ -91,18 +92,48 @@ function displayTopResultsForCurrentTrend(countryCode) {
...
+              snippetArray.forEach((word) => {
+                // Replace quotations or other symbols that may be attached to 
+                // the word.  
+                let checkWordInArray = word.replace(/[^a-z0-9+]+/gi, '');
+                if (results[i].score > 0 && sentimentWordsResult.positive
+                    .includes(checkWordInArray)) {
+                  resultElement.innerHTML += '<span style=\'color: ' + 
+                      POSITIVE_COLOR + ';\'>' + word + ' </span>';
+                } else if (results[i].score < 0 && sentimentWordsResult.negative
+                    .includes(checkWordInArray)) {
```
Comment (carmenBen) [inline search-sentiments-map/public/scripts/modal-script.js:119]: Have you considered looping through the positive or negative words rather than every word in the snippet? Because then you only loop through the words you know need to change.

File: search-sentiments-map/public/scripts/modal-script.js [hunk 15]
```diff
@@ -91,18 +92,48 @@ function displayTopResultsForCurrentTrend(countryCode) {
...
+                  resultElement.innerHTML += '<span style=\'color: ' + 
+                      NEGATIVE_COLOR + ';\'>' + word + ' </span>';
+                } else {
+                  resultElement.innerHTML += word + ' ';
```
Comment (carmenBen) [inline search-sentiments-map/public/scripts/modal-script.js:123]: Wait if you always add a space at the end, does that mean there is a space between the last word and the final punctuation?

File: search-sentiments-map/public/scripts/modal-script.js (hunk 6 above)
Comment (carmenBen) [reply search-sentiments-map/public/scripts/modal-script.js:127]: +1

File: search-sentiments-map/public/scripts/modal-script.js [hunk 16]
```diff
@@ -91,18 +92,48 @@ function displayTopResultsForCurrentTrend(countryCode) {
...
+            } else {
+              resultElement.innerHTML += results[i].snippet + '<br>' + 
+                  '<i>Sentiment Score: ' + results[i].score.toFixed(1) + 
+                  '</i><br>';
+            }
+          });
     }
+    resultElement.innerHTML += '<i>Last updated on ' + date + '</i>';
   }
-  resultElement.innerHTML += '<i>Last updated on ' + date + '</i>';
```
Comment (carmenBen) [inline search-sentiments-map/public/scripts/modal-script.js:105]: Why did you move this into the else? Even if there are no results, we still want to show when it was last updated.

File: search-sentiments-map/routes/sentiment-words.js [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+  let titleSnippet = req.params.titleSnippet;
+  getPositiveNegativeWords(titleSnippet).then(sentimentWordsJsonArray => {
+    res.setHeader('Content-Type', 'application/json');
+    res.send(sentimentWordsJsonArray);
+  });
+});
+
+async function getPositiveNegativeWords(titleSnippet) {
+  try {
+    var result = await sentimentWords.analyze(titleSnippet);
```
Comment (carmenBen) [inline search-sentiments-map/routes/sentiment-words.js:33]:
```suggestion
    let result = await sentimentWords.analyze(titleSnippet);
```

File: search-sentiments-map/routes/sentiment-words.js (hunk 9 above)
Comment (carmenBen) [reply search-sentiments-map/routes/sentiment-words.js:41]: +1

File: search-sentiments-map/public/scripts/modal-script.js (hunk 2 above)
Comment (ntarn) [reply search-sentiments-map/public/scripts/modal-script.js:78]: It is 60 lines long, but I think since all the code is relevant to the function title, and we do have other 60 lines long functions. I think it'd be more clear to just have the same code in there bc it is relevant to the function instead of sending it to another smaller function. What do others think? @hayleejane3 @carmenBen @lawrencezjh

File: search-sentiments-map/package.json (hunk 10 above)
Comment (ntarn) [reply search-sentiments-map/package.json:17]: Then that would be a trailing comma

File: search-sentiments-map/public/scripts/modal-script.js (hunk 14 above)
Comment (ntarn) [reply search-sentiments-map/public/scripts/modal-script.js:119]: Yea I thought about that possibility before coding it, but I don't think there's then a way to singly color a word in the snippet without going through each word, bc each word could be a different color

Comment (ChenyuHeidiZhang) [reply search-sentiments-map/public/scripts/modal-script.js:119]: Is there any javascript function for find and replace elements in an array so that you can find the words that you look for and replace them with their colored version

File: search-sentiments-map/public/scripts/modal-script.js (hunk 2 above)
Comment (ChenyuHeidiZhang) [reply search-sentiments-map/public/scripts/modal-script.js:78]: Which other function is 60 lines long? This function has too many layers of logic and I think you can put a coherent subsection in a different function.

File: search-sentiments-map/public/scripts/modal-script.js (hunk 14 above)
Comment (carmenBen) [reply search-sentiments-map/public/scripts/modal-script.js:119]: Or could you split the string at that word and then concatenate the strings back together with the new colored word?

Comment (ChenyuHeidiZhang) [reply search-sentiments-map/public/scripts/modal-script.js:119]: yeah that's also what I mean

File: search-sentiments-map/public/scripts/modal-script.js (hunk 2 above)
Comment (hayleejane3) [reply search-sentiments-map/public/scripts/modal-script.js:78]: The name is completely fine. I've seen functions whose names don't even fit on a single line!
Descriptiveness is good in this case.

File: search-sentiments-map/public/scripts/modal-script.js [hunk 18]
```diff
@@ -91,18 +92,48 @@ function displayTopResultsForCurrentTrend(countryCode) {
...
-          results[i].snippet + '<br>' + '<i>Sentiment Score: ' + 
-          results[i].score.toFixed(1) + '</i><br>';
+      await fetch('/sentiment-words/' + results[i].title + results[i].snippet)
+          .then(resultsJsonArray => resultsJsonArray.json())
+          .then(sentimentWordsResult => {
+            // i + 1 shows the index for each search result.
+            resultElement.innerHTML += (i + 1).toString() + '. ' + '<a href=' + 
+                results[i].link + '>' + results[i].htmlTitle + '</a><br>';
+            // Check to make sure that there is detection of positive and 
+            // negative words from the Node.js sentiment API.
```
Comment (hayleejane3) [inline search-sentiments-map/public/scripts/modal-script.js:106]: Prefer active voice over passive, and to the point :)

```suggestion
            // Check that positive and negative words are detected by the Node.js sentiment API.
```

File: search-sentiments-map/public/scripts/modal-script.js (hunk 14 above)
Comment (hayleejane3) [reply search-sentiments-map/public/scripts/modal-script.js:119]: +1 to the easier approaches using existing util functions.

File: search-sentiments-map/public/scripts/modal-script.js (hunk 2 above)
Comment (ChenyuHeidiZhang) [reply search-sentiments-map/public/scripts/modal-script.js:78]: @hayleejane3 We are talking about the length of the actual function, not the name

Comment (ntarn) [reply search-sentiments-map/public/scripts/modal-script.js:91]: retrieveUserSearchResultFromDatastore in search.js is 50+ lines, I just think the layers of logic makes most sense to be left in this since this function is supposed to be updating resultElement.innerHTML.

Comment (ntarn) [reply search-sentiments-map/public/scripts/modal-script.js:91]: made it shorter with the new logic

Comment (hayleejane3) [reply search-sentiments-map/public/scripts/modal-script.js:91]: my bad

File: search-sentiments-map/public/scripts/modal-script.js [hunk 19]
```diff
@@ -105,50 +106,71 @@ async function displayTopResultsForCurrentTrend(countryCode) {
...
+      let positiveHighlight = '<span style="color: ' +
+            POSITIVE_COLOR + ';">' + snippet.substring(wordIndex, 
+            wordIndex + word.length) + '</span>';
+      let negativeHighlight = '<span style="color: ' +
+            NEGATIVE_COLOR + ';">' + snippet.substring(wordIndex, 
+            wordIndex + word.length) + '</span>';
+      snippet = scoreIsPositive ? snippet.replace(new RegExp(
+          '\\b'+ word + '\\b', 'gi'), positiveHighlight) : 
+          snippet.replace(new RegExp('\\b' + word + '\\b', 'gi'), 
+          negativeHighlight);
```
Comment (hayleejane3) [inline search-sentiments-map/public/scripts/modal-script.js:169]: I think it should be something like this

@ChenyuHeidiZhang could you double check too?

```suggestion
            POSITIVE_COLOR + ';">' +
            snippet.substring(wordIndex, wordIndex + word.length) +
            '</span>';
      let negativeHighlight = '<span style="color: ' +
            NEGATIVE_COLOR + ';">' +
            snippet.substring(wordIndex,  wordIndex + word.length) +
            '</span>';
      snippet = scoreIsPositive ?
          snippet.replace(new RegExp( '\\b'+ word + '\\b', 'gi'), positiveHighlight) : 
          snippet.replace(new RegExp('\\b' + word + '\\b', 'gi'), negativeHighlight);
```

File: search-sentiments-map/public/scripts/modal-script.js [hunk 20]
```diff
@@ -105,50 +106,71 @@ async function displayTopResultsForCurrentTrend(countryCode) {
...
+            // sentiment API.
+            if (sentimentWordsResult != null) {
+              let positiveWords = sentimentWordsResult.positive;
+              let negativeWords = sentimentWordsResult.negative;
+              if (results[i].score > 0 && positiveWords.length !== 0) {
+                highlightWords(snippet, new Set(positiveWords), true, 
+                    resultElement);
+              } else if (results[i].score < 0 && negativeWords.length !== 0) {
+                highlightWords(snippet, new Set(negativeWords), false, 
+                    resultElement);
```
Comment (hayleejane3) [inline search-sentiments-map/public/scripts/modal-script.js:131]:
```suggestion
                highlightWords(
                    snippet, new Set(positiveWords), true, resultElement);
              } else if (results[i].score < 0 && negativeWords.length !== 0) {
                highlightWords(
                    snippet, new Set(negativeWords), false, resultElement);
```

File: search-sentiments-map/public/scripts/modal-script.js (hunk 19 above)
Comment (ntarn) [reply search-sentiments-map/public/scripts/modal-script.js:169]: I just checked, and you're correct, I knew we had to break highest syntactically, but I wasn't sure what to do then because the line wasn't broken at those syntatic points. I just looked at the example again, and it's similar to what you do here. Thanks!
https://google.github.io/styleguide/jsguide.html#formatting-where-to-break

File: search-sentiments-map/routes/sentiment-words.js [hunk 21]
```diff
@@ -20,22 +20,14 @@ const express = require('express');
...
-    res.send(sentimentWordsJsonArray);
-  });
-});
-
-async function getPositiveNegativeWords(titleSnippet) {
+router.get('/:snippet', (req, res) => {
   try {
-    var result = await sentimentWords.analyze(titleSnippet);
-    return result;
+    var sentimentWordsJsonArray = sentimentWords.analyze(req.params.snippet);
```
Comment (carmenBen) [inline search-sentiments-map/routes/sentiment-words.js:25]: Set to use `let` instead of `var`

File: search-sentiments-map/public/scripts/modal-script.js [hunk 22]
```diff
@@ -105,50 +106,72 @@ async function displayTopResultsForCurrentTrend(countryCode) {
...
+ * @param {boolean} scoreIsPositive Boolean for whether or not the sentiment 
+ *     score is positive.
+ * @param {Object} resultElement Tab element to update with the search result 
+ *     snippet.
+ */
+function highlightWords(
+    snippet, wordsToHighlight, scoreIsPositive, resultElement) {
+  let startDisplayingSnippetIndex = 0;
+  wordsToHighlight.forEach((word) => {
+    let wordIndex = snippet.toLowerCase().indexOf(word);
```
Comment (carmenBen) [inline search-sentiments-map/public/scripts/modal-script.js:158]: Add a comment explaining why you only need to transform the snippet to lower case, and not the word too.

File: search-sentiments-map/public/scripts/modal-script.js [hunk 23]
```diff
@@ -105,50 +106,72 @@ async function displayTopResultsForCurrentTrend(countryCode) {
...
+          POSITIVE_COLOR + ';">' +
+          snippet.substring(wordIndex, wordIndex + word.length) +
+          '</span>';
+      let negativeHighlight = '<span style="color: ' +
+          NEGATIVE_COLOR + ';">' +
+          snippet.substring(wordIndex,  wordIndex + word.length) +
+          '</span>';
+      snippet = scoreIsPositive ?
+          snippet.replace(new RegExp( '\\b'+ word + '\\b', 'gi'), positiveHighlight) : 
+          snippet.replace(new RegExp('\\b' + word + '\\b', 'gi'), negativeHighlight);
```
Comment (carmenBen) [inline search-sentiments-map/public/scripts/modal-script.js:170]:
```suggestion
      let highlight = '<span style="color: ' +
          (scoreIsPositive ? POSITIVE_COLOR : NEGATIVE_COLOR) +
          ';">' + snippet.substring(wordIndex,  wordIndex + word.length) +
          '</span>';
      snippet.replace(new RegExp('\\b' + word + '\\b', 'gi'), highlight);
```
